apm-sniffer/apm-sdk-plugin/jdbc-commons/src/main/java/org/apache/skywalking/apm/plugin/jdbc/define/StatementEnhanceInfos.java

```diff
@@ -31,11 +34,13 @@
     private ConnectionInfo connectionInfo;
     private String statementName;
     private String sql;
+    private Map<Integer, Object> parametersKeyedByIndex;
 
     public StatementEnhanceInfos(ConnectionInfo connectionInfo, String sql, String statementName) {
         this.connectionInfo = connectionInfo;
         this.sql = sql;
         this.statementName = statementName;
+        this.parametersKeyedByIndex = new TreeMap<Integer, Object>();
```

Comment:
1. Why treeMap?
2. Don't create the map if no parameter tracing. Map creation is a high payload. You could test 😄

Reply:
Treemap to keep order? User may not using the 1-n order. You need to bw careful.

Reply:
1. We need to keep the parameters' order, because user may invoke `ps.setString(2, "val2");` before `ps.setString(1, "val1")`, and we need to print `val2, val1`.
2. Got it, will improve this later

Reply:
> We need to keep the parameters' order, because user may invoke ps.setString(2, "val2"); before ps.setString(1, "val1"), and we need to print val2, val1.

No, we don't. Think in this way, dev or ops team doesn't know the codes, and saw `select xx from yyy where a=? and b=?` and you put `val2(i = 1), val1(i = 2)`, how they think?

Reply:
I think people thought `a = val2 and b = val1`, right? But the reality is `a = val1 and b = val2` right?
Tracing is not really all about codes, it is for observability, a.k.a. understandable for the system.

Reply:
> > We need to keep the parameters' order, because user may invoke ps.setString(2, "val2"); before ps.setString(1, "val1"), and we need to print val2, val1.
> 
> No, we don't. Think in this way, dev or ops team doesn't know the codes, and saw `select xx from yyy where a=? and b=?` and you put `val2(i = 1), val1(i = 2)`, how they think?

by saying "keep the parameters' order", I didn't mean to keep the **order of the code**, I meant keep the **index order of the code**, and since the `index` is the key of the `TreeSet`, therefore the values are sorted by the keys (index here),

Reply:
Then why don't use an Array list? Look like much more efficiency, right?
 ```
list = new ArrayList;
list,checkRange(i)
list.set(i, value)
```

Reply:
we don't know the parameters' size in advance, `checkRange` manually may need to enlarge the capacity, and copying elements, I'm not sure which one is better, did I miss anything about `ArrayList`?

Reply:
First of all, even in today's agent codes, we have an unsolved issue. In many practices, introduced by Google, ArrayList has better performance than LinkedList, even with enlarge the capacity and copy elements. We can build a benchmark. Clearly, LinkedList should similar or better than TreeMap in complexity, right? Also, the benchmark is welcome.

Besides above, let's think in this way. ArrayList with 20 as initial size, then how many of parameters a SQL will have? 40 - 60 is a big one, right? Especially in micro-service and less Oracle today.
The ArrayList will enlarge twice mostly, others just set, I think that should be much quicker.

Let's do multiple tests to confirm this.

Reply:
make sense, I'll try to do some benchmark tests to make this clear and for future reference

Reply:
Benchmark and result are here: https://github.com/apache/skywalking/pull/2846/files#diff-c79b0eaced528877516f9cbf5a4ddb7c